File: lambda/metrics.py
"""Custom metrics module for CloudWatch."""

import json
import time
from typing import Dict


class CustomMetrics:
    """
    Custom metrics emission using CloudWatch Embedded Metric Format (EMF).
    This allows us to emit custom metrics directly from logs without additional API calls.
    """
    
    @staticmethod
    def emit_metric(metric_name: str, value: float = 1.0, unit: str = "Count", 
                   namespace: str = "BitbucketTeamsWebhook", dimensions: Dict[str, str] = None):
        """
        Emit a custom metric using CloudWatch Embedded Metric Format.
        
        Args:
            metric_name: Name of the metric
            value: Metric value (default 1.0 for counters)
            unit: Metric unit (Count, Milliseconds, etc.)
            namespace: CloudWatch namespace
            dimensions: Additional dimensions for the metric
        """
        if dimensions is None:
            dimensions = {}
        
        # Create EMF log entry
        emf_log = {
            "_aws": {
                "Timestamp": int(time.time() * 1000),  # Milliseconds since epoch
                "CloudWatchMetrics": [
                    {
                        "Namespace": namespace,
                        "Dimensions": [list(dimensions.keys())] if dimensions else [[]],
                        "Metrics": [
                            {
                                "Name": metric_name,
                                "Unit": unit
                            }
                        ]
                    }
                ]
            },
            metric_name: value
        }
        
        # Add dimensions to the log entry
        emf_log.update(dimensions)
        
        # Emit as structured log (CloudWatch will parse this automatically)
        print(json.dumps(emf_log))
    
    @staticmethod
    def emit_event_type_metric(event_type: str):
        """Emit metric for specific event type"""
        CustomMetrics.emit_metric(
            f"EventType-{event_type.replace(':', '-')}",
            namespace="BitbucketTeamsWebhook/EventTypes",
            dimensions={"EventType": event_type}
        )
    
    @staticmethod
    def emit_signature_failure():
        """Emit metric for signature verification failure"""
        CustomMetrics.emit_metric("SignatureVerificationFailures")
    
    @staticmethod
    def emit_teams_api_failure():
        """Emit metric for Teams API failure"""
        CustomMetrics.emit_metric("TeamsAPIFailures")
    
    @staticmethod
    def emit_unsupported_event():
        """Emit metric for unsupported event type"""
        CustomMetrics.emit_metric("UnsupportedEventTypes")
    
    @staticmethod
    def emit_processing_duration(duration_ms: float):
        """Emit metric for processing duration"""
        CustomMetrics.emit_metric(
            "ProcessingDuration",
            value=duration_ms,
            unit="Milliseconds"
        )

File: src/aws/metrics.ts
/**
 * Metrics Emitter Module
 * Emits custom metrics to CloudWatch using EMF (Embedded Metric Format)
 * Requirements: 11.1, 11.2, 11.3, 11.4, 11.5, 11.6
 */

const NAMESPACE = 'BitbucketTeamsWebhook';

interface EMFMetric {
  _aws: {
    Timestamp: number;
    CloudWatchMetrics: Array<{
      Namespace: string;
      Dimensions: string[][];
      Metrics: Array<{ Name: string; Unit: string }>;
    }>;
  };
  [key: string]: any;
}

function buildEMFMetric(
  metricName: string,
  value: number = 1,
  unit: string = 'Count',
  dimensions?: Record<string, string>
): EMFMetric {
  const dimensionEntries = dimensions ? Object.entries(dimensions) : [];
  const dimensionNames = dimensionEntries.map(([key]) => key);

  const metric: EMFMetric = {
    _aws: {
      Timestamp: Date.now(),
      CloudWatchMetrics: [
        {
          Namespace: NAMESPACE,
          Dimensions: dimensionNames.length > 0 ? [dimensionNames] : [[]],
          Metrics: [{ Name: metricName, Unit: unit }]
        }
      ]
    },
    [metricName]: value
  };

  dimensionEntries.forEach(([key, val]) => {
    metric[key] = val;
  });

  return metric;
}

export function emitMetric(
  metricName: string,
  value: number = 1,
  unit: string = 'Count',
  _namespace?: string,
  dimensions?: Record<string, string>
): void {
  console.log(JSON.stringify(buildEMFMetric(metricName, value, unit, dimensions)));
}

export const emitEventTypeMetric = (eventType: string) =>
  emitMetric(`EventType-${eventType}`, 1, 'Count', NAMESPACE, { EventType: eventType });

export const emitSignatureFailure = () =>
  emitMetric('SignatureVerificationFailures', 1, 'Count', NAMESPACE, {
    ErrorType: 'SignatureVerificationFailure'
  });

export const emitTeamsAPIFailure = () =>
  emitMetric('TeamsAPIFailures', 1, 'Count', NAMESPACE, { ErrorType: 'TeamsAPIFailure' });

export const emitUnsupportedEvent = () =>
  emitMetric('UnsupportedEventTypes', 1, 'Count', NAMESPACE, { ErrorType: 'UnsupportedEventType' });

export const emitProcessingDuration = (durationMs: number) =>
  emitMetric('ProcessingDuration', durationMs, 'Milliseconds', NAMESPACE, {
    MetricType: 'ProcessingDuration'
  });
